modal/pi0/app.py

```diff
@@ -0,0 +1,663 @@
+    dataset_name: str,
+    wandb_api_key: str | None,
+    model_name: str,
+    training_params: TrainingParamsPi05,
+    user_hf_token: str | None = None,
+    private_mode: bool = False,
+    timeout_seconds: int = FUNCTION_TIMEOUT_TRAINING,
+    **kwargs,
+):
+    """
+    Pi0 training function.
+
+    We run LoRA fine tuning on the provided dataset and upload the trained model to HuggingFace.
+    Optionnal: If a wandb_api_key is provided, we log the training to Weights & Biases.
+    """
+    from datetime import datetime, timezone
+    from supabase import Client, create_client
+
+    SUPABASE_URL = os.environ["SUPABASE_URL"]
+    SUPABASE_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
+    supabase_client: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
+
+    hf_token = user_hf_token or os.getenv("HF_TOKEN")
+    if hf_token is None:
+        raise ValueError(
+            "HF_TOKEN is not available (neither user token nor system token)"
+        )
+
+    logger.info(
+        f"🚀 Training pi0.5 on {dataset_name} with id {training_id} and uploading to: {model_name}  (private_mode={private_mode})"
+    )
+    wandb_run_url: str | None = None
+    training_params_dict = training_params.model_dump(
+        exclude_none=True,
+    )
+    config_name = training_params_dict.get("config_name", "pi0.5_LoRA_finetune_so100")
+    exp_name = training_params_dict.get("exp_name", "pi05_so100_lora_finetune")
+
+    try:
+        # Update training status to running
+        supabase_client.table("trainings").update(
+            {
+                "status": "running",
+                "requested_at": datetime.now(timezone.utc).isoformat(),
+            }
+        ).eq("id", training_id).execute()
+
+        wandb_enabled = wandb_api_key is not None
+
+        if wandb_enabled:
+            try:
+                wandb.login(key=wandb_api_key, verify=True)
+            except Exception as e:
+                logger.info(
+                    f"Failed to login to Weights & Biases: {e}. Disabling Weights & Biases."
+                )
+                wandb_enabled = False
+
+        # Start by computing normalization stats
+
+        norm_cmd = [
+            "uv",
+            "run",
+            "/openpi-source/scripts/compute_norm_stats.py",
+            config_name,
+            "--data.repo_id=" + dataset_name,
+        ]
+
+        logger.info(f"Training parameters: {training_params_dict}")
+        for key, value in training_params_dict.items():
+            norm_cmd.append(f"--{key}={value}")
+
+        # Run the command
+        logger.info(f"Computing normalization stats with command: {' '.join(norm_cmd)}")
+
+        run(norm_cmd, check=True)
+        logger.info("Normalization stats computed successfully")
+        # Then run the training
+
+        train_cmd = [
+            "uv",
+            "run",
+            "/openpi-source/scripts/train.py",
+            config_name,
+            "--data.repo_id=" + dataset_name,
+        ]
```

Comment:
can we instead directly call the main() function from the script ? this way we'd get autocompletion, type safety, and better tracebacks